worker/peergrouper/worker_test.go

```diff
@@ -125,64 +125,83 @@ func (s *workerSuite) waitAdvance(c *gc.C, d time.Duration) {
 
 		logger.Infof("starting worker")
 		s.newNoPublishWorker(c, st)
+		// Due to the inherit complexity of the multiple goroutines running
+		// and listen do different watchers, there is no way to manually
+		// advance the testing clock in a controlled manner as the clock.After
+		// calls can be replaced in response to other watcher events. Hence
+		// using the standard testing clock wait / advance method does not
+		// work. So we use the real clock to advance the test clock for this
+		// test.
+		done := make(chan struct{})
+		go func() {
+			for {
+				select {
+				case <-time.After(5 * time.Millisecond):
+					s.clock.Advance(pollInterval)
+				case <-done:
+					return
+				}
+			}
+		}()
 
 		// Wait for the worker to set the initial members.
 		mustNext(c, memberWatcher)
 		assertMembers(c, memberWatcher.Value(), mkMembers("0v 1 2", ipVersion))
 
 		// Update the status of the new members
 		// and check that they become voting.
-		c.Logf("updating new member status")
-		s.waitAdvance(c, pollInterval)
+		c.Logf("\nupdating new member status")
 		st.session.setStatus(mkStatuses("0p 1s 2s", ipVersion))
 		mustNext(c, memberWatcher)
 		assertMembers(c, memberWatcher.Value(), mkMembers("0v 1v 2v", ipVersion))
 
-		c.Logf("adding another machine")
-		// Add another machine.
+		c.Logf("\nadding another machine")
 		m13 := st.addMachine("13", false)
 		m13.setStateHostPort(fmt.Sprintf(ipVersion.formatHostPort, 13, mongoPort))
 		st.setControllers("10", "11", "12", "13")
 
-		c.Logf("waiting for new member to be added")
+		c.Logf("\nwaiting for new member to be added")
 		mustNext(c, memberWatcher)
 		assertMembers(c, memberWatcher.Value(), mkMembers("0v 1v 2v 3", ipVersion))
 
-		// Remove vote from an existing member;
-		// and give it to the new machine.
-		// Also set the status of the new machine to
-		// healthy.
-		c.Logf("removing vote from machine 10 and adding it to machine 13")
+		// Remove vote from an existing member; and give it to the new
+		// machine. Also set the status of the new machine to healthy.
+		c.Logf("\nremoving vote from machine 10 and adding it to machine 13")
 		st.machine("10").setWantsVote(false)
 		st.machine("13").setWantsVote(true)
 
 		st.session.setStatus(mkStatuses("0p 1s 2s 3s", ipVersion))
 
 		// Check that the new machine gets the vote and the
 		// old machine loses it.
-		c.Logf("waiting for vote switch")
+		c.Logf("\nwaiting for vote switch")
 		mustNext(c, memberWatcher)
 		assertMembers(c, memberWatcher.Value(), mkMembers("0 1v 2v 3v", ipVersion))
 
-		c.Logf("removing old machine")
+		c.Logf("\nremoving old machine")
 		// Remove the old machine.
 		st.removeMachine("10")
 		st.setControllers("11", "12", "13")
 
 		// Check that it's removed from the members.
-		c.Logf("waiting for removal")
+		c.Logf("\nwaiting for removal")
 		mustNext(c, memberWatcher)
 		assertMembers(c, memberWatcher.Value(), mkMembers("1v 2v 3v", ipVersion))
+		// Stop the clock advancing goroutine.
+		close(done)
```

Comment:
This should be a defer above or else the clock advancing goroutine ends up being left behind if the test fails.